### PR TITLE
add KEP-377: proposal about Karmada ResourceQuota

### LINKED ISSUE #377: support Resource Quotas in karmada
**What would you like to be added**:
support Resource Quotas in karmada

**Why is this needed**:
In our scenario, several users or teams will share the ‘cluster’, many users can use all the resources provided. If there is no restriction, someone will be hungry. In order to make all users use resources fairly, we need to support Resource Quotas

### PR DESCRIPTION
**Pull request by aven-ai**
Signed-off-by: as1919@qq.com

**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
Add a new proposal, KEP-377:  proposal about Karmada ResourceQuota

**Which issue(s) this PR fixes**:
Fixes #377 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
proposal about Karmada ResourceQuota
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
@aven-ai Thanks for the awesome proposal!!!
Can you add some information about the background of the proposal?

File: docs/proposals/resource-quota/README.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+ - Provide complex quotas to meet various business scenarios.
+
+## Proposal
+
+### Use cases
+
+1. Ability to enumerate resource usage limits per namespace.
+2. Ability to monitor resource usage for tracked resources.
+3. Ability to reject resource usage exceeding hard quotas.
```
[the remark this answers is not in the document]
Comment (aven-ai) [reply docs/proposals/resource-quota/README.md:40]: Emm, sorry. I preset a premise that people are already familiar with and used ResourceQuota of Kubernetes. Strictly speaking, the description of Ability to... is a brief overview of the function, not a user story. If a user wants to find the usage scenarios and benefits of ResourceQuota on Karmada, we should have this introduction. Let me change and add it.

**Comment by aven-ai** [comment]
> @aven-ai Thanks for the awesome proposal!!!
> Can you add some information about the background of the proposal?

Of course, thank you for your suggestions.

File: docs/proposals/resource-quota/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply docs/proposals/resource-quota/README.md:40]: > I preset a premise that people are already familiar with and used ResourceQuota of Kubernetes.

I didn't mean to add descriptions to ResourceQuota. 

The `ResourceQuota` is a K8S resource that can be propagated by defining `PropagationPolicy`, I guess your proposal wants to address something else. A user story would be easier to show the benefits.

[the remark this answers is not in the document]
Comment (aven-ai) [reply docs/proposals/resource-quota/README.md:47]: > > I preset a premise that people are already familiar with and used ResourceQuota of Kubernetes.
> 
> I didn't mean to add descriptions to ResourceQuota.
> 
> The `ResourceQuota` is a K8S resource that can be propagated by defining `PropagationPolicy`, I guess your proposal wants to address something else. A user story would be easier to show the benefits.

As we discussed, I have added some background and user stories about the global quota, in hope that users can have a more comprehensive understanding of this feature. Thank you for your exchange.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply docs/proposals/resource-quota/README.md:47]: Thanks. Looks better now.

**Comment by RainbowMango** [comment]
cc @kevin-wangzefeng Please take a look.

**Comment by kevin-wangzefeng** [comment]
1. It's not clear to me whether this proposal is trying to propose an new API or just enable the K8s ResourceQuota API cheking in Karmada control plane. If it's the former one, I'd suggest use a new API kind as well as clarify the API group.

2. To enable the global resource quota restriction, two things need to be enforced:

	2.1. quota consumption checking in the karmada control plane: reject any requests (to karmada apiserver) that would exceed the global quota during API creation 

	2.2. coordinate with the quota checking in member clusters: e.g. pod creation by deployments in member cluster
	* If strong consistency is required, we will need to hack the quota checking process in member clusters, calling back to a "global quota checking service" to make sure no exceedance happen.
	* If only eventual consistency is required, we can dynamically divide global quota and adjust quota settings in member cluster accordingly. The API call checking process in member clusters keep as usual.

**Comment by RainbowMango** [comment]
Hi @aven-ai ,
I see you just made a new push. Are you gonna show the update at the next community meeting?

**Comment by aven-ai** [comment]
> 1. It's not clear to me whether this proposal is trying to propose an new API or just enable the K8s ResourceQuota API cheking in Karmada control plane. If it's the former one, I'd suggest using a new API kind as well as clarify the API group.
> 2. To enable the global resource quota restriction, two things need to be enforced:
>    2.1. quota consumption checking in the karmada control plane: reject any requests (to karmada apiserver) that would exceed the global quota during API creation
>    2.2. coordinate with the quota checking in member clusters: e.g. pod creation by deployments in member cluster
>    
>    * If strict consistency is required, we will need to hack the quota checking process in member clusters, calling back to a "global quota checking service" to make sure no exceedance happen.
>    * If only eventual consistency is required, we can dynamically divide global quota and adjust quota settings in member cluster accordingly. The API call checking process in member clusters keep as usual.

Last time we discussed the Quota of karmada, thanks for your exchange, I have some additions about these

1. Karmada's Quota is not just enable the K8s ResourceQuota API cheking in Karmada control plane, it plans to propose a new API. In order to avoid ambiguity with ResourceQuota, we plan to use **`KarmadaQuota`** to name this Quota proposal. I updated the proposal.

2. KarmadaQuota is designed as a central Quota, which controls the object's request resources such as Deployment and Statfulset by monitoring ResourceBinding. KarmadaQuota's watch for ResourceBinding is equivalent to the pod of ResourceQuota.

3. If a pod created in sub-clusters, karmada is not able to perceive them, and then KarmadaQuota will not limit them too. In addition, if a pod that Karmada can not perceive is created in a sub-cluster, this using will destroy the management of multiple clusters by Karmada. I think this using should be avoided.

4. HPA scenario, if the HPA of the sub-cluster itself is used, the replica result of the sub-cluster is inconsistent with the karmada scheduling result. It may cause karmada rescheduling(KarmadaQuota can recalculate during the reschedule), so  The design of karmada HPA should also use centralized management.

**Comment by aven-ai** [comment]
> Hi @aven-ai ,
> I see you just made a new push. Are you gonna show the update at the next community meeting?

Yes, I have made some explained about my design ideas in my reply to kevin, we can make a disscuss on next meeting. also we can disscuss more in this proposal.

**Comment by YouHang** [comment]
What if a customer use karmada to propagate a controller which can create pods in the member clusters? As the quota webhook only check the used resources at propagation phase in the federation cluster, these pod will be created in the member clusters lead to a customer can use resources more than the quota limit.

**Comment by aven-ai** [comment]
> What if a customer use karmada to propagate a controller which can create pods in the member clusters? As the quota webhook only check the used resources at propagation phase in the federation cluster, these pod will be created in the member clusters lead to a customer can use resources more than the quota limit.

In this case, the user bypassed Karmada's resource scheduling and distribution, and secretly created pods in the member cluster. These pods cannot be perceived by Karmada, and KarmadaQuota cannot take effect on these pods too.
This situation actually damages karmada's management and control of multi-cluster resources, which is equivalent to use resources privately in a single cluster，we do not recommend this usage. Thanks for your point of view, I will remind this scenario's risk for KarmadaQuota in the proposal.

**Comment by RainbowMango** [comment]
Hi @aven-ai 
I can't trigger the test(don't know why it hanging), So please rebase and squash your commits.
Let's merge it first then iterate it on this version.

**Comment by aven-ai** [comment]
> I see you just made a new

**Comment by aven-ai** [comment]
> Hi @aven-ai I can't trigger the test(don't know why it hanging), So please rebase and squash your commits. Let's merge it first then iterate it on this version.

ok

**Comment by likakuli** [comment]
>Resourcebinding of karmada is designed to only support ResourceRequest, and the limit of pod Container does not be distributed to the sub-cluster, so KarmadaQuota only supports Request type resources.

Is there a plan of `KarmadaQuota` to support Limit type resources? I met the global quota management requirement, but what we need is to verify the resource limit but not request.

**Comment by RainbowMango** [comment]
@likakuli can you open an issue to explain your requirement in detail?

**Comment by likakuli** [comment]
> @likakuli can you open an issue to explain your requirement in detail?
of course

**Comment by likakuli** [comment]
https://github.com/karmada-io/karmada/issues/1927

**Pull request merged**